core/sr-io/src/offchain/http.rs

```diff
@@ -0,0 +1,398 @@
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		fmt.debug_struct("ResponseBody")
+			.field("id", &self.id)
+			.field("is_deadline_reached", &self.is_deadline_reached)
+			.field("buffer", &self.buffer.len())
+			.field("filled_up_to", &self.filled_up_to)
+			.field("position", &self.position)
+			.field("deadline", &self.deadline)
+			.finish()
+	}
+}
+
+impl ResponseBody {
+	fn new(id: RequestId) -> Self {
+		ResponseBody {
+			id,
+			is_deadline_reached: false,
+			buffer: [0_u8; 4096],
+			filled_up_to: None,
+			position: 0,
+			deadline: None,
+		}
+	}
+
+	/// Set the deadline for reading the body.
+	pub fn deadline(&mut self, deadline: impl Into<Option<Timestamp>>) {
+		self.deadline = deadline.into();
+	}
+
+	/// Check if the iterator has finished because of reaching the deadline.
+	pub fn is_deadline_reached(&self) -> bool {
+		self.is_deadline_reached
+	}
+}
+
+impl Iterator for ResponseBody {
+	type Item = u8;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		if self.is_deadline_reached {
+			return None;
+		}
+
+		if self.filled_up_to.is_none() {
+			let result = crate::http_response_read_body(self.id, &mut self.buffer, self.deadline).ok();
+			if let Some(size) = result {
+				self.position = 0;
+				self.filled_up_to = Some(size);
+			} else {
+				self.is_deadline_reached = true;
+				return None;
+			}
+		}
+
+		if Some(self.position) == self.filled_up_to {
+			self.filled_up_to = None;
+			return self.next();
+		}
+
+		let result = self.buffer[self.position];
+		self.position += 1;
+		Some(result)
+	}
+}
+
+/// A collection of Headers in the response.
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Headers {
+	/// Raw headers
+	pub raw: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+impl Headers {
+	/// Retrieve a single header from the list of headers.
+	///
+	/// Note this method is linearly looking from all the headers.
+	/// If you want to consume multiple headers it's better to iterate
+	/// and collect them on your own.
+	pub fn find(&self, name: &str) -> Option<&str> {
+		for &(ref key, ref val) in &self.raw {
+			if from_utf8(&key) == Some(name) {
+				return from_utf8(&val)
+			}
+		}
+		None
+	}
+
+	/// Convert this headers into an iterator.
+	pub fn into_iter(&self) -> HeadersIterator {
+		HeadersIterator { collection: &self.raw, index: None }
+	}
+}
+
+/// A custom iterator traversing all the headers.
+#[derive(Clone)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct HeadersIterator<'a> {
+	collection: &'a [(Vec<u8>, Vec<u8>)],
+	index: Option<usize>,
+}
+
+impl<'a> HeadersIterator<'a> {
+	/// Move the iterator to the next position.
+	///
+	/// Returns `true` is `current` has been set by this call.
+	pub fn next(&mut self) -> bool {
+		let index = self.index.map(|x| x + 1).unwrap_or(0);
+		self.index = Some(index);
+		index < self.collection.len()
+	}
+
+	/// Returns current element (if any).
+	///
+	/// Note that you have to call `next` prior to calling this
```

Comment:
Maybe if `next` wasn't called instead of `self.index.unwrap_or(0)` it would be better to return `None`?

I wonder if `self.index?` is possible : )